nemo/collections/asr/modules/audio_modules.py

```diff
@@ -0,0 +1,686 @@
+
+    @property
+    def input_types(self) -> Dict[str, NeuralType]:
+        """Returns definitions of module output ports.
+        """
+        return {
+            "input": NeuralType(('B', 'C', 'D', 'T'), SpectrogramType()),
+            "input_length": NeuralType(('B',), LengthsType()),
+        }
+
+    @property
+    def output_types(self) -> Dict[str, NeuralType]:
+        """Returns definitions of module output ports.
+        """
+        return {
+            "output": NeuralType(('B', 'C', 'D', 'T'), FloatType()),
+            "output_length": NeuralType(('B',), LengthsType()),
+        }
+
+    @typecheck()
+    def forward(self, input: torch.Tensor, input_length: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Estimate `num_outputs` masks from the input spectrogram.
+
+        Args:
+            input: C-channel input, shape (B, C, F, N)
+            input_length: Length of valid entries along the time dimension, shape (B,)
+
+        Returns:
+            Returns `num_outputs` masks in a tensor, shape (B, num_outputs, F, N),
+            and output length with shape (B,)
+        """
+        input, _ = self.features(input=input, input_length=input_length)
+        B, num_feature_channels, num_features, N = input.shape
+
+        # (B, num_feat_channels, num_feat, N) -> (B, N, num_feat_channels, num_feat)
+        input = input.permute(0, 3, 1, 2)
+
+        # (B, N, num_feat_channels, num_feat) -> (B, N, num_feat_channels * num_features)
+        input = input.view(B, N, -1)
+
+        # Apply projection on num_feat
+        input = self.input_projection(input)
+
+        # Apply RNN on the input sequence
+        input_packed = torch.nn.utils.rnn.pack_padded_sequence(
+            input, input_length.cpu(), batch_first=True, enforce_sorted=False
+        ).to(input.device)
+        self.rnn.flatten_parameters()
+        input_packed, _ = self.rnn(input_packed)
+        input, input_length = torch.nn.utils.rnn.pad_packed_sequence(input_packed, batch_first=True)
+        input_length = input_length.to(input.device)
+
+        # Create `num_outputs` masks
+        output = []
+        for output_projection in self.output_projections:
+            # Output projection
+            mask = output_projection(input)
+            mask = self.output_nonlinearity(mask)
+
+            # Back to the original format
+            # (B, N, F) -> (B, F, N)
+            mask = mask.transpose(2, 1)
+
+            # Append to the output
+            output.append(mask)
+
+        # Stack along channel dimension to get (B, M, F, N)
+        output = torch.stack(output, axis=1)
+
+        # Mask frames beyond input length
+        length_mask: torch.Tensor = make_seq_mask_like(
+            lengths=input_length, like=output, time_dim=-1, valid_ones=False
+        )
+        output = output.masked_fill(length_mask, 0.0)
+
+        return output, input_length
+
+
+class MaskReferenceChannel(NeuralModule):
```

Comment:
The above two class names are a bit odd. In the neural type hierarchy - they are Mask type operations yes but refer to how we define ASR encoder and decoder classes. 
I suppose this is fine for now, but IMO it just looks a bit odd.

Reply:
resolved offline.